### PR TITLE
fix relative/absolute path handling for external FBX textures

### PR DESCRIPTION
**Pull request by birarda**
- fixed a bug loading external textures at relative paths

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: d4112f0e3691f2a7a56ea3124083050e0c928b74 available at: https://gist.github.com/f4bca7d379f09b4008bdd50e79d30131<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10116-d4112f0.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10116-d4112f0.dmg

File: libraries/fbx/src/FBXReader.cpp [hunk 1]
```diff
@@ -833,11 +833,10 @@ FBXGeometry* FBXReader::extractFBXGeometry(const QVariantHash& mapping, const QS
                         const int MODEL_UV_SCALING_MIN_SIZE = 2;
                         const int CROPPING_MIN_SIZE = 4;
                         if (subobject.name == "RelativeFilename" && subobject.properties.length() >= RELATIVE_FILENAME_MIN_SIZE) {
-                            QByteArray filename = subobject.properties.at(0).toByteArray();
-                            QByteArray filepath = filename.replace('\\', '/');
-                            filename = fileOnUrl(filepath, url);
```
Comment (zzmp) [inline libraries/fbx/src/FBXReader.cpp]: Does the fileOnUrl function still need to be declared, or can it be removed?

Comment (birarda) [inline libraries/fbx/src/FBXReader.cpp]: You're right - assuming you agree with this change I can pull it.

**Comment by zzmp** [comment]
Can you elaborate on the bug that this is fixing?

**Comment by birarda** [comment]
@zzmp I'll add a test plan that should make it clear

**Comment by hifi-gustavo** [comment]
Build console output for commit: b9783d768eecb64011385bf9aa74fd014c19eabd available at: https://gist.github.com/f4bca7d379f09b4008bdd50e79d30131<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10116-b9783d7.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10116-b9783d7.dmg

**Comment by zzmp** [comment]
This looks good, I just want to see a test plan so I know what it's trying to fix.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4e07a6a865b36f583b551b581bfc2a02163ba5dd available at: https://gist.github.com/f4bca7d379f09b4008bdd50e79d30131<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10116-4e07a6a.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10116-4e07a6a.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 137e73cf210e6e74f5ef9c349fb75d526b18439d available at: https://gist.github.com/f4bca7d379f09b4008bdd50e79d30131<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10116-137e73c.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10116-137e73c.dmg

**Comment by birarda** [comment]
#### Test Plan

Must test all possible texture references for FBX, for local and remote files (ATP/HTTP).

To load an FBX as a local file use a path like `file:///C:/test/cube.fbx`.

##### Embedded Texture

This FBX has an embedded texture. Test loading it as a local file, from HTTP, and from ATP. 

In each case, the cube should render as expected, with the texture.
In each case, you should see a line like the following in your log:

##### External Texture by Absolute Path

This FBX has an external texture that it references using an absolute path. Test loading it as a local file, from HTTP, and from ATP.

##### External Texture by Relative Path

http://hifi-public.s3.amazonaws.com/birarda/models/DiscGolfBasket.ktx.fbx

This FBX uses external textures that it references at a relative path, in a directory `textures` beside the FBX. Test loading it as a local file, from HTTP, and from ATP. 

Note that this FBX file will attempt to load textures that are not present. The model with render in Interface without textures. For this part of the test you only need to confirm that the URLs shown in the log are correct.

###### Local File

Your log should have lines like the following (your local path may be different):

```
[04/04 15:19:29] [DEBUG] [hifi.networking] Finished loading: file:///C:/cygwin64/home/birarda/code/lod/DiscGolfBasket.ktx.fbx
[04/04 15:19:29] [DEBUG] [hifi.networking] Error loading  QUrl("file:///C:/cygwin64/home/birarda/code/lod/textures/DiscGolfBasket_Albedo.ktx")
[04/04 15:19:29] [DEBUG] [hifi.networking] Failed to load: file:///C:/cygwin64/home/birarda/code/lod/textures/DiscGolfBasket_Albedo.ktx
```

On my machine, the FBX is at `file:///C:/cygwin64/home/birarda/code/lod/DiscGolfBasket.ktx.fbx`. Notice that the texture load is correctly attempted at `file:///C:/cygwin64/home/birarda/code/lod/textures/DiscGolfBasket_Albedo.ktx`.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 70eada0f76a8ff829b64f952a6d1af1f0a628435 available at: https://gist.github.com/f4bca7d379f09b4008bdd50e79d30131<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10116-70eada0.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10116-70eada0.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 137e73cf210e6e74f5ef9c349fb75d526b18439d available at: https://gist.github.com/f4bca7d379f09b4008bdd50e79d30131<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10116-137e73c.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10116-137e73c.dmg

**Comment by birarda** [comment]
@zzmp I'm waiting on a model for the "External Texture by Absolute Path" test case - but the final test case of "External Texture by Relative Path" is what was not working before and is fixed by this PR.

**Comment by 68chase** [comment]
Testing

**Comment by birarda** [comment]
@68chase I just pushed a build, would you mind waiting until that one shows up to test with that?

**Comment by hifi-gustavo** [comment]
Build console output for commit: e4021c168e7ef32b27d6cf2c17f2ef22bcf7cd99 available at: https://gist.github.com/f4bca7d379f09b4008bdd50e79d30131<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10116-e4021c1.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10116-e4021c1.dmg

**Comment by 68chase** [comment]
@birarda No Problem.  When should I check back?

**Comment by birarda** [comment]
@68chase it just popped up! The URL is http://cachefly.highfidelity.com/HighFidelity-Beta-PR10116-e4021c1.exe.

**Comment by 68chase** [comment]
@birarda Got it

**Comment by 68chase** [comment]
@birarda Need some clarification please.  I am on the second step where I have to load the model by HTTP.  Does that just mean to load the asset by using Import Entities.  The dialog box (see attached) says select Models to Import, but the .fbx is not showing.   I apologize if its a newb question, but I am still pretty new.
Thanks

image

**Comment by birarda** [comment]
@68chase no problem! You should be able to click "Edit" in your tablet/toolbar and then click "Model". That should prompt you for a model URL where you can add the HTTP URL. Let me know if that works.

**Comment by 68chase** [comment]
@birarda Perfect, thank you.  Ill let you know

**Comment by 68chase** [comment]
Hey @birarda  Just FYI - I know this is easy to do.   The model CoinPile_V2.fbx is spelled two different ways in the document (Sometime with a **V**2 and sometimes with a **v**2).  I was copying and pasting your instructions and I couldn't get it to work.   Just wanted to let you know in case you do the same thing.  It passed though. Ill post my pics in a minute

**Comment by 68chase** [comment]
Passed



3 http load
4 http load
5 local files system non embedded textures
6 http load non embedded textures
7 local files system non embedded textures
8 http non embedded textures
9 atp loaded
10 final test filesystem
11 final test http
12 final test atp 1
13 final test atp 2

**Comment by OmegaHeron** [comment]
This seems to have introduced a change to how existing avatar assets are loaded.  I'm noticing issue loading up my avatar since this landed.  Whether this is a bug (to me it *is*) or a new intended behavior is up to debate.

Previously I would see an http transaction like so;
`ip_redacted - - [10/Apr/2017:16:01:22 -0400] "GET /assets/av/tex/mega-wip-4-clothed_Top_Normal.png HTTP/1.1" 200 360513 "-" "Mozilla/5.0 (HighFidelityInterface)"`

With this PR same FST attempts to load;
`ip_redacted - - [10/Apr/2017:15:59:55 -0400] "GET /assets/av/tex/mega-wip-2/orig/mega-wip-4-clothed/mega-wip-4-clothed_Top_Normal.png HTTP/1.1" 404 152 "-" "Mozilla/5.0 (HighFidelityInterface)"`

Master requests asset from `/assets/av/tex/`
While PR requests for same FST from `/assets/av/tex/mega-wip-2/orig/mega-wip-4-clothed/`

That's likely path info being yanked from FBX that was discarded/ignored before.

Here's what matters in FST.
`name = mega4clothedv4-pbr`
`type = body+head`
`scale = 1`
`filename = av/mega-pbr-5.fbx`
`texdir = av/tex`

So we're good for fbx source as it loads model at /av/mega-pbr-5.fbx then it seems to concatenate some cruft path data left in FBX + texdir from FST.

**Comment by birarda** [comment]
@OmegaHeron I'm looking into this now - can you send me the FBX that is being loaded differently?

It is the case that an FBX with relative paths to textures would previously only load those textures right beside the FBX and is now loading them at that relative path.

**Comment by OmegaHeron** [comment]
Will ping you in slack @birarda

**Comment by OmegaHeron** [comment]
Also - for FBX, right - it expected textures to be in same relative dir as FBX.  FST worked differently... it ignored everything/anything else and loaded textures from texdir defined line.  Now it's adding data to texdir path and borking things.  Note critical difference between FST behavior and FBX prior to this PR.  So this is an issue with FST, not so much FBX.

**Pull request merged**